src/Catalog/Monitoring/Validation/Test/ValidationTestRunner.cs

```diff
@@ -0,0 +1,109 @@
+using System.Threading.Tasks;
+using NuGet.Packaging.Core;
+using System.Threading;
+using NuGet.Services.Metadata.Catalog.Monitoring.Validation.Result;
+using NuGet.Services.Metadata.Catalog.Monitoring.Providers;
+
+namespace NuGet.Services.Metadata.Catalog.Monitoring.Validation.Test
+{
+    /// <summary>
+    /// Performs a validation test on a <see cref="ValidationTestRunData"/>.
+    /// </summary>
+    public interface IValidationTestRunner
+    {
+        /// <summary>
+        /// Returns whether or not the validation test should run on a <see cref="ValidationTestRunData"/>.
+        /// </summary>
+        Task<bool> ShouldRun(ValidationTestRunData data);
+
+        /// <summary>
+        /// Validates a package.
+        /// </summary>
+        /// <returns>A <see cref="ValidationTestResult"/> which contains the results of the validation.</returns>
+        Task<ValidationTestResult> Run(ValidationTestRunData data);
+    }
+    
+    /// <summary>
+    /// Abstract class with the shared functionality between all <see cref="IValidationTestRunner"/> implementations.
+    /// </summary>
+    public abstract class ValidationTestRunner : IValidationTestRunner
+    {
+        /// <summary>
+        /// The number of tries attempted before the test is considered to have failed.
+        /// </summary>
+        public const int MaxTries = 3;
+
+        protected ValidationTestRunner(
+            ResourceToTest<IPackageTimestampMetadataResource, V2ResourceType> timestampMetadataResourceV2)
+        {
+            TimestampMetadataResourceV2 = timestampMetadataResourceV2?.Resource ??
+                throw new ArgumentNullException(nameof(timestampMetadataResourceV2));
+        }
+
+        protected readonly IPackageTimestampMetadataResource TimestampMetadataResourceV2;
+
+        /// <summary>
+        /// Checks that the current batch of catalog entries contains the entry that was created from the current state of the V2 feed.
+        /// If the metadata is missing from either V2 or the catalog, validation runs anyway.
+        /// </summary>
+        public virtual async Task<bool> ShouldRun(ValidationTestRunData data)
+        {
+            var timestampV2 = await TimestampMetadataResourceV2.Get(data);
+            var timestampCatalog = await PackageTimestampMetadata.FromCatalogEntries(data.Client, data.Entries);
+
+            // A couple notes here:
+            // 1 - If the timestamp metadata in the catalog is LESS than that of the feed, we must not be looking at the latest entry that corresponds with this package, so skip the test for now.
+            // 2 - If the timestamp metadata in the catalog is EQUAL to that of the feed, we are looking at the latest catalog entry that corresponds with this package, so run the test.
+            // 3 - If the timestamp metadata in the catalog is GREATER than that of the feed, there is likely some issue--the catalog should never have newer data than the feed--so run the test.
```

Comment:
>timestamp [](start = 26, length = 9)

Looks like case 3 and 4 are exceptional. Why continue as normal? Can't we just fail and force investigation into weirdness?